Feature/friendly url dist patch1

GodsDev opened the pull request:
Core code:
* MyFriendlyURL is a new routing part of Controller (of MVC) (Examples of behaviour explained in dist/README.md#seo are implemented in dist)
* Controller is set by Controller::prepareTemplate a Controller::prepareTemplateAll and called by MyController::run (instead of MyController::controller, which still remains for backward compatibility)

Dist code:
* rozsáhlejší úpravy dist, aby se rovnou nainstalovalo funkční včetně datové struktury stránek a produktů vč. friendly URL směrování ve 4 jazycích (cs,de,en,fr) a užití redirektoru

Governance and Coding stype:
* README.md and dist/README.md: rozsáhlé doplnění komentářů, aby bylo srozumitelnější; 
* PHPUnit test: friendly URL testováno
* Coding style: https://www.php-fig.org/psr/psr-12/ replaces PSR-2
* Coding style: array() to [] (As of PHP 5.4)
* Coding style: removal of spaces (PHP CS Fixer)
* Many elements ordered alphabetically for better readability
* Removal of unused code

New methods:
* MyCommon::verboseBarDump - Dumps information about a variable in Tracy Debug Bar or is silent
* MyController::redir - Redirects to $redir (incl. relative) and die
* MyController::run - spuštění, které používá Controller::prepareTemplate a Controller::prepareTemplateAll
* ProjectCommon::getLinkSql - Returns SQL fragment for column link ($fieldName) which construct either parametric URL or relative friendly URL
* ProjectCommon::language - Returns or set the language
* MyTableAdmin: fix if prefill value is not yet among the existing values, set as value for the own-value input box

View:
* latte may use $applicationDirLanguage to keep the selected language folder
* FEATURE_FLAGS are available now in javascript

WorkOfStan committed:
composer.json: support links added
added ProjectCommon->language() Sets language (instead of having it in every project specific ProjectSpecific)
added ProjectCommon->getLinkSql() Returns SQL fragment for column link ($fieldName) which construct either parametric URL or relative friendly URL (instead of having it in every project specific ProjectSpecific)
added MyCommon->verboseBarDump($var, $title = null, array $options = []) Dumps information about a variable in Tracy Debug Bar or dumps it to standard output
MyFriendlyUrl - better described

dist/.htaccess: If mod_alias is not enabled, the Apache should fail to show anything in order to always hide the files in the section below
dist/.htaccess: Friendly URL processing enabled
dist/build.sh: composer set to autoload classes from the classmap only. Implicitly enables `--optimize-autoloader`.
dist/conf/config.php: default values for 'FRIENDLY_URL', 'HOME_TOKEN', 'FORCE_301', 'REDIRECTOR_ENABLED' - explained in `dist/README.md`
dist/index.php: Controller called with `'requestUri' => $_SERVER['REQUEST_URI']` and request parameters POST are also taken into account
dist/classes/FriendlyUrl.php: new

WorkOfStan committed:
simplify MyCommon->verboseBarDump($var) Dumps information about a variable in Tracy Debug Bar or is silent based on `$this->verbose` set when instantiating Controller

WorkOfStan committed:
comment Texy related

WorkOfStan committed:
array() to []

WorkOfStan committed:
const TEMPLATE_NOT_FOUND, TEMPLATE_DEFAULT moved to MyCommon as they MUST be the same both for Controller extends MyController extends MyCommon and MyFriendlyUrl extends MyCommon
FriendlyUrl->switchParametric() contains example to be implement in database
MyController->redir() moved from Controller->redir() projects

MyController->run()
* Determines template, set Session language, runs prepareTemplate for single template and prepareAllTemplates for general transformations
* Outputs changed $MyCMS->template and $MyCMS->context as fields of an array
* replaces ->controller() /which still remains for backward compatibility/

dist/conf/config.php
* description of RULES for switchParametric to be configured in 'templateAssignementParametricRules'
* $WEBSITE['menu']: populates the default menu in `header.latte`, key means URL

`dist/conf/config.local.dist.php` - more legible

WorkOfStan committed:
fix prepareTemplate and prepareAllTemplates MUST be defined in MyController even though the methods of same name but from the child class will only be used
example latte created so that the menu links work

WorkOfStan committed:
cosmetic and removal of obsoleted code

WorkOfStan committed:
changed to protected so it can be inherited and therefore used in its proper context

WorkOfStan committed:
cosmetic: obsolete code removal and PSR-12 spaces and new lines

WorkOfStan committed:
removing unused folder sql

WorkOfStan committed:
removal of unused settings

WorkOfStan committed:
* Description of friendly URL related constants
* Constants put in alphabetical order for better legibility

WorkOfStan committed:
* as friendlyUrl populates $this->MyCMS->context['applicationDir'] it is cleaner to use {$applicationDir} in .latte instead of {$_SERVER['SCRIPT_NAME'])}

WorkOfStan committed:
* $controller->friendlyUrl instantiated in context of project (dist)

WorkOfStan committed:
Friendly URL mechanism explained

WorkOfStan committed:
* better description of language management and friendly URL/redirect calculation
* add languages fr and de
* change cn to zh for Chinese according to ISO 639-1 https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes

WorkOfStan committed:
* SEO Force 301 language folder

WorkOfStan committed:
* SEO Force 301 language folder changed to 302 language folder (this kind of redirect may be caused by using the same token for multiple languages without using language subpatern folder, so 302 Found aka Moved Temporarily is the best choice)
* de and fr translations
* parse path regexp pattern is calculated from array_keys($MyCMS->TRANSLATIONS) without DEFAULT_LANGUAGE to match interestingPath with `language subpattern` and the `rest of the path` in the method MyFriendlyUrl::friendlyIdentifyRedirect

WorkOfStan committed:
* add database table_prefix field to phinx.yml
* init_admin_table migration written using proper object methods
* create redirector_table

WorkOfStan committed:
Default content structure in database

WorkOfStan committed:
* admin - display row options described
* fix admin: if prefill value is not yet among the existing values, set as value for the own-value input box
* fix admin: if multiple column values needed in label - use array of columns instead of trying to use SQL statements such as CONCAT

WorkOfStan committed:
* admin: update Summernote to v.0.8.18 (2020-05-20) - hence HTML editor icons are visible

WorkOfStan committed:
* URL fields have btn-webalize button

WorkOfStan committed:
better description

GodsDev committed:
Create product.latte

GodsDev committed:
product and category routing prepared

* category 1 part of menu

GodsDev committed:
Create category.latte

GodsDev committed:
Create list.latte

GodsDev committed:
prepare template product and category

WorkOfStan committed:
* new product(s) latte

WorkOfStan committed:
fix news to list

WorkOfStan committed:
Merge branch 'feature/FriendlyUrlDist' into feature/FriendlyUrlDist-patch1

WorkOfStan committed:
* category is addressed as ?category&id=N
* list.latte: idGlue &id= or =

WorkOfStan committed:
better debugging

WorkOfStan committed:
better description and logging

WorkOfStan committed:
* category visualisation
* product visualisation
* ProjectSpecific: update getCategory, update getContent, create getProduct
* obsolete Controller->sectionStyles removed
* new (unfinished) FriendlyUrlTest.php

WorkOfStan committed:
* fix: the FriendlyURL translation to parametric URL is taken into account
* optimisation: pureFriendlyUrl recursion only through the loop of 'templateAssignementParametricRules' instead of the whole determineTemplate()
* dist/classes/FriendlyUrl.php::findFriendlyUrlToken adapted to mycmsprojectnamespace

WorkOfStan committed:
fix: product=1 has friendlyUrl `alfa`

WorkOfStan committed:
* fetchSingle exception if error
* latte may use $applicationDirLanguage to keep the selected language
* hide product language variants with empty title both in list and 404 if called directly
* README.md accepted types of multilingual URLs

WorkOfStan committed:
this->verbose may be loosely equal to true to invoke Debugger::barDump

WorkOfStan committed:
removal of spaces (PHP CS Fixer) and of unused code

WorkOfStan committed:
* removal of spaces (PHP CS Fixer)
* removal of unused code
* refactoring: getting rid of variable MyController::redir::$sqlStatementsArray, MyCMSMonoLingual::renderLatte::$sqlStatementsArray

WorkOfStan committed:
page titles for demo pages

WorkOfStan committed:
FriendlyUrlTest::testPageStatusOverHttp to test webserver results of Friendly and Parametric URLs

WorkOfStan committed:
Product samples added to database

WorkOfStan committed:
* SEO related redirects of product&id=1 to something tested and updated in README.md
* refactoring FriendlyUrl::findFriendlyUrlToken
* default category has friendly url_XX

WorkOfStan committed:
refactor prepareTemplateAll method for category option populating $this->MyCMS->context['list']

WorkOfStan committed:
* article contact created
* description: how to add new content
* typeToTableMapping for type uses specific table for its records set in config.php instead of FriendlyUrl::findFriendlyUrlToken
* findFriendlyUrlToken moved to MyFriendlyUrl so if the default behaviour is ok, it doesn't have to be overidden
* product context expanded from json to array

WorkOfStan committed:
* fix persistent language with parametric URL
* removal of unused code
* finishing FriendlyUrlTest including Redirector tests old_url => new_url

WorkOfStan committed:
phpunit test from root uses the same language as from dist/, i.e. `cs`

WorkOfStan committed:
coding standard: PHP CS fixer

WorkOfStan committed:
Home is now part of all language variants of menu

WorkOfStan committed:
Removal of unused latte templates

WorkOfStan committed:
PHP CS fixer

WorkOfStan committed:
* newer robmorgan/phinx
* latest commit of godsdev/mycms (TODO: after merge to develop and master - change to ^0.4.0)

WorkOfStan committed:
Fix MD034/no-bare-urls Bare URL used

WorkOfStan committed:
* fix ProjectCommon::$requestUri is right (instead of ProjectSpecific::$requestUri)
* fix dist/.htaccess not to switch default-something to de language
* fix ?article&code=contacts work even without FRIENDLY_URL == true
* fix new Controller::projectSpecific moved after Controller->language has been properly set (i.e. fixing scenario when after /?language=cs the /en/alfa still had czech menu
* fix autotrack.2.4.1.js sourceMappingURL=autotrack.2.4.1.js.map
* FEATURE_FLAGS are available now in javascript
* @layout.latte block script now might include change of variables expected in index.js below

WorkOfStan committed:
* fix MD041/first-line-heading/first-line-h1
* github workflows/php-composer-dependencies.yml

WorkOfStan committed:
* fix getSessionLanguage - strip back slashes to normalize both xx and xx/ to xx
* readability: dist/.htaccess - alphabetic order

WorkOfStan committed:
removal of obsolete code

WorkOfStan committed:
* adding use \Nette\SmartObject
* a bit different coding style concerning blank lines so that https://github.com/github/super-linter likes it

GodsDev committed:
Code standardization (#37)

Code governance
* phpunit is only require-dev, so `webmozart/assert` MUST be required in the `composer.json` of this application
* Check composer --no-dev dependencies
* build.sh set Executable bit (+x) to fix Warnings found in [bash-exec] linter! [WARN ] Error: File:[/github/workspace/build.sh] is not executable
* CustomFilters::webalize renamed because Method name "CustomFilters::webalize_" is not in camel caps format
* context['article'] to context['content'] for article.latte
* .gitignore: .php_cs.cache
* PHPUnit: error_reporting(E_ALL); // incl E_NOTICE
* change of author email rejthar@stanislavrejthar.com
* bump version "godsdev/tools": "^0.3.7
* bump version backyard/3.2.9 ("minimum-stability": "stable")
* gitignore cache
* forgotten HOME_TOKEN settings hinted by PHPUnit
* phplint.yml uses actions/checkout@v2 version as well
* main branch (will replace `master` branch in due time)
* dist/build.sh: To work on low performing environments, the script accepts number of seconds as parameter to be used as a waiting time between steps.
* PHP/5.6 tested and PHP/7.2 tested

Code changes and various minor bugs fixes
* TableAdmin.php: Localised field content\\url_## has language independant behaviour
* TableAdmin.php: TRANSLATIONS are defined through constructor, uses web app config
* LogMysqli: int instead of bool Parameter #2 of method GodsDev\MyCMS\LogMysqli::query() so that it is compatible with method GodsDev\Backyard\BackyardMysqli::query()
* MyAdmin: definition of properties $ASSETS_SUBFOLDERS, $agendas instead of their definition in dist/classes/Admin.php
* MyAdmin: Method GodsDev\MyCMS\MyAdmin::projectSpecificSections() invoked with 1 parameter, 0 required.
* MyAdminProcess: fix type Tools::addMessage($result['success'], $result['messages']); (instead of message)
* MyAdminProcess: $salt = mt_rand((int) 1e8, (int) 1e9); (Fix: mt_rand expects int, float given)
* MyTableAdmin: removed __construct as it only duplicated parent
* MyTableLister::viewInputs - $option typo changed to $options
* fix MyCMSMonoLingual::csrfStart: Parameter #1 $min and #2 $max of function rand expects int, float given.
* fix MyController Access to an undefined properties GodsDev\MyCMS\MyController::$language and GodsDev\MyCMS\MyController::$requestUri.
* fix MyCMSMonoLingual Result of (!is_object($this->logger) || !($this->logger instanceof LoggerInterface)) is always false.
* fix type for MyAdminProcess::$tableAdmin
* fix MyTableLister::setTable throw new \RunTimeException
* fix TableAdmin: LogMysqli database object instead of vanilla \mysqli
* LogMysqli::query does not throw \DBQueryException as it logs troubles instead
* LogMysqli::query @return \mysqli_result Object|false
* MyTableLister::viewTable argument \mysqli_result $query (instead of \mysqli)
* TableAdmin::customSave - typecast to fix parameter #1 $input of function str_pad expects string, int given.
* AdminProcess::adminProcess - fix Binary operation "+" between string and (float|int) results in an error. Parameter #1 $input of function str_pad expects string, (float|int) given.
* config.php: fix Parameter #2 $newvalue of function ini_set expects string, int given.
* fix MyFriendlyUrl::PAGE_NOT_FOUND - Visibility must be declared on all constants if your project supports PHP 7.1 or later
* fix MyFriendlyUrl - The first trait import statement must be declared on the first non-comment line after the class opening brace
* fix MyAdminProcess::processFilePack - Parentheses must be used when instantiating a new class (\ZipArchive)
* Parentheses must be used when instantiating a new class: MyCMSMonoLingual::renderLatte, MyAdminProcess::processFileUnpack

Code style
* PHP linter and super-linter for GitHub Actions
* $this->tableAdmin everywhere instead of $this->TableAdmin or $TableAdmin
* markdown lint README.md: MD010/no-hard-tabs, MD009/no-trailing-spaces, MD014/commands-show-output, MD025/single-title/single-h1, MD040/fenced-code-language, MD009/no-trailing-spaces, MD037/no-space-in-emphasis
* VALIDATE_CSS: false (stylelint seems to evaluate sass the same way as CSS; `.stylelintignore` prepared for future use)

PHPStan
* recommended fixes to PHPDocs
* removal of wrong code
* phpstan include structure (mycms includes dist, github/phpstan includes local; when PHPStan test dist/ as part of MyCMS, it knows the GodsDev\\MyCMS\\* classes, so they should be ignored only if dist/ lives in its own project) 
* ignoreErrors references to 3rd library objects
* reasonable ignoreErrors exceptions
* dynamicConstantNames instead of constant induced exceptions
* some mixed type hints replaced with more specific ones
* discovering-symbols, e.g. conf/constants.php: Predefining constants for PHPSTAN analysis as recommanded by <https://phpstan.org/user-guide/discovering-symbols> 

PHPCS
* 120 chars limit enforced
* multi-line control structure
* loosen rules where it makes sense

TODO fix (CRS2)
* MyTableAdmin: the right correction is replacing #3 $cols and #4 $rows false,false by 60,5 as int is expected?
* MyTableLister: addPage() invoked with 6 parameters, 3-5 required; Parameter #1 $types of method GodsDev\MyCMS\MyTableLister::filterKeys() expects array, string given

Authored-by: Stanislav Rejthar, rejthar@stanislavrejthar.com

GodsDev commented on dist/classes/AdminProcess.php:
```diff
@@ -104,32 +99,58 @@ public function adminProcess(&$post)
+            isset($post['product-switch'], $post['id']) &&
+            ($product = $this->MyCMS->dbms->query(
+                // TODO 3x Expected at least 1 space after "+"; 0 found ask CRS2
+                'SELECT category_id,sort FROM ' . TAB_PREFIX . 'product WHERE id=' . +$post['id']
```
@crs2 co dělá to + před $post['id']? Má to být int casting?

crs2 replied:
Ano. (Tahle odpověď platí i pro ostatní výskyty.)

GodsDev commented on dist/classes/AdminProcess.php:
```diff
@@ -104,32 +99,58 @@ public function adminProcess(&$post)
+            )->fetch_assoc())
+        ) {
+            $id = $this->MyCMS->fetchSingle('SELECT id FROM ' . TAB_PREFIX . 'product WHERE category_id='
+                . +$product['category_id'] . ' AND sort' . ($post['product-switch'] == 1 ? '>' : '<')
```
@crs2 co dělá to + před $product['category_id']? Má to být int casting?

GodsDev commented on dist/classes/AdminProcess.php:
```diff
@@ -104,32 +99,58 @@ public function adminProcess(&$post)
+                $this->MyCMS->dbms->query('UPDATE ' . TAB_PREFIX . 'product SET sort='
+                    . $product['sort'] . ' WHERE id=' . $id);
+                $this->MyCMS->dbms->query('UPDATE ' . TAB_PREFIX . 'product SET sort='
+                    . ($product['sort'] + $post['product-switch']) . ' WHERE id=' . +$post['id']);
```
@crs2 co dělá to + před $post['id']? Má to být int casting?

GodsDev commented on dist/classes/AdminProcess.php:
```diff
@@ -183,6 +208,7 @@ public function adminProcess(&$post)
         // delete a table record
         if (isset($post['record-delete'])) {
+            // TODO nemá náhodou být ->customAfterDelete namísto customDelete? Ask crs2
             if (!$this->tableAdmin->customDelete()) {
```
@crs2 nemá to náhodou být customAfterDelete ? Neboť customDelete není.

crs2 replied:
Upřímně, nevím. V dist/classes/TableAdmin.php je customDelete().

GodsDev replied:
Aha - proměnná tableAdmin měla zděděný typ MyTableAdmin. Doplnil jsem tedy, že tableAdmin je typu TableAdmin. Slávek

GodsDev commented on classes/MyAdmin.php:
```diff
@@ -408,25 +421,26 @@ protected function outputBodyEnd()
             . 'EXPAND_INFIX = "' . EXPAND_INFIX . '";' . PHP_EOL
             . 'TOKEN = ' . end($_SESSION['token']) . ';' . PHP_EOL
+            // todo CRS2 fix Parameter #2 $options of function json_encode expects int, true given.
             . 'ASSETS_SUBFOLDERS = ' . json_encode($this->ASSETS_SUBFOLDERS, true) . ';' . PHP_EOL
```
@crs2 co má být jako druhý parametr? Je tam true, ale má tam když tak být int.

crs2 replied:
Nechť je tam json_encode($this->ASSETS_SUBFOLDERS)

GodsDev commented on classes/MyAdmin.php:
```diff
@@ -631,79 +642,97 @@ protected function outputTableEditSelected($clone = false)
                     if (Tools::ends($value['type'], 'text') || Tools::ends($value['type'], 'blob')) {
-                        $result .= Tools::htmlTextarea("fields[$key]", '', null, 3, ['class' => 'form-control edit-selected', 'data-size' => $value['size']]) . '</td>';
+                        // TODO ask CRS2 if replacing #3 $cols null by 60 as int is expected is the right correction
+                        $result .= Tools::htmlTextarea("fields[$key]", '', 60, 3, ['class' => 'form-control edit-selected', 'data-size' => $value['size']]) . '</td>';
```
@crs2 upravil jsem v htmlTextarea parametr 3 z null na 60, což je defaultní hodnota té metody. Je to správná úprava?

GodsDev commented on classes/MyAdminProcess.php:
```diff
@@ -144,11 +151,14 @@ public function processClone(&$post)
+                    exit; //@todo
                 }
+                // TODO nedosažitelný kód - ask CRS2
                 if ($sql) {
```
@crs2 K tomuhle kódu se nelze dostat, Buď je ten zbytečný ten kód anebo ty exity před tím?

GodsDev commented on classes/MyAdminProcess.php:
```diff
@@ -176,6 +186,7 @@ public function processExport(&$post, $get)
                     $sql = $sql['select'];
                 } else { //export only checked rows
+                    // TODO ask CRS2 - filterToSql doesn't exists. $errors isn't defined. $sql isn't defined.
                     $this->filterToSQL($get, $sql, $errors);
```
@crs2 co ten kód má dělat? Metoda filterToSql neexistuje. $errors ani $sql nejsou definovány.

GodsDev commented on classes/MyAdminProcess.php:
```diff
@@ -489,6 +498,7 @@ public function processSubfolder(&$post)
                 if ($post['info'] && class_exists('\ZipArchive')) {
                     $ZipArchive = new \ZipArchive();
+                    // todo ask CRS2 if this block should encapsulate also the following foreach or if otherwise it should fail with exception
                 }
```
@crs2 Nemá náhodou být součástí bloku té podmínky `if ($post['info'] && class_exists('\ZipArchive'))` také následující foreach blok, kde se počítá s výsledkem ZipArchive? Anebo tam má být else blok, kde bude Exception, že ZipArchive nezafungovalo?

GodsDev commented on classes/MyAdminProcess.php:
```diff
@@ -560,10 +570,16 @@ public function processUserChangePassword(&$post)
                     if ($row['active'] == '1' && $row['password_hashed'] == sha1($post['old-password'] . $row['salt'])) {
-                        Tools::resolve($this->MyCMS->dbms->query('UPDATE ' . TAB_PREFIX . 'admin
+                        // TODO ask CRS2 Static method GodsDev\Tools\Tools::resolve() invoked with 5 parameters, 3 required.
+                        Tools::resolve(
```
@crs2 Jaký je záměr 4. a 5. parametru, když metoda resolve počítá jen se 3 parametry?

GodsDev commented on classes/MyTableAdmin.php:
```diff
@@ -205,11 +206,12 @@ protected function outputField(array $field, $key, array $record, array $options
             } else {
-                $output .= Tools::htmlTextarea("fields[$key]", $value, false, false, [
+                // TODO ask CRS2 if replacing #3 $cols and #4 $rows false,false by 60,5 as int is expected is the right correction
+                $output .= Tools::htmlTextarea("fields[$key]", $value, 60, 5, [
```
@crs2 3. a 4. parametr metody htmlTextarea byly false, false, ale metoda očekává int, tak jsem to změnil na defaultní hodnoty metody, tedy 60,5 - je to správně?

GodsDev commented on classes/MyTableAdmin.php:
```diff
@@ -306,10 +331,17 @@ protected function outputField(array $field, $key, array $record, array $options
-                    ])
+                $input = '<div class="TableAdminTextarea">'
+                    // TODO ask CRS2 if replacing #3 $cols and #4 $rows false,false by 60,5 as int is expected is the right correction
+                    . Tools::htmlTextarea(
```
@crs2 3. a 4. parametr metody htmlTextarea byly false, false, ale metoda očekává int, tak jsem to změnil na defaultní hodnoty metody, tedy 60,5 - je to správně?

GodsDev commented on classes/MyTableAdmin.php:
```diff
@@ -283,16 +303,21 @@ protected function outputField(array $field, $key, array $record, array $options
-                                'label-class' => 'font-weight-normal'
+                            'type' => 'checkbox',
+                            // todo ask CRS2 is_array($value) seeems to be always true, so Else branch is unreachable because ternary operator condition is always true.
+                            'checked' => ((1 << $k) & (int) (is_array($value) ? reset($value) : $value)) ? 'checked' : null,
```
is_array($value) je false pokud $value před switch bylo '' (anebo false). Možná by šlo refaktorovat.

GodsDev commented on classes/MyTableAdmin.php:
```diff
@@ -454,6 +487,7 @@ public function recordSave($messageSuccess = false, $messageError = false)
             }
             foreach ($this->fields as $key => $field) {
+                // todo ask CRS2 - Variable $value might not be defined.
                 if (Tools::set($_POST['fields-null'][$key]) || (Tools::set($field['foreign_table']) && $value === '')) {
```
@crs2 $value nemusí zde být vůbec ještě definováno. Co v ní má být? Poslední hodnota v poli $_POST?

crs2 replied:
Místo && $value === '' má být && $field === ''.

GodsDev commented on classes/MyTableAdmin.php:
```diff
@@ -476,27 +510,31 @@ public function recordSave($messageSuccess = false, $messageError = false)
                 } else {
-                    $where .= ' AND ' . (is_null($original) ? Tools::escapeDbIdentifier($key) . ' IS NULL' : ($original . '' === '' ? 'IFNULL(' . Tools::escapeDbIdentifier($key) . ', "")' : Tools::escapeDbIdentifier($key)) .' = "' . $this->escapeSQL($_POST['original'][$key]) . '"');
+                    // todo ask CRS2 Variable $original might not be defined.
+                    $where .= ' AND ' . (is_null($original) ? Tools::escapeDbIdentifier($key) . ' IS NULL' : ($original . '' === '' ? 'IFNULL(' . Tools::escapeDbIdentifier($key) . ', "")' : Tools::escapeDbIdentifier($key)) . ' = "' . $this->escapeSQL($_POST['original'][$key]) . '"');
```
@crs2 Pokud $this->fields nebylo pole, tak $original nebylo vůbec definováno. Jak to ošetřit?

crs2 replied:
Zkusme změnit druhou větev if v recordSave() z:
if (is_array($this->fields)) {
na:
if (count($this->fields) > 0) {
— $this->fields má být pole s alespoň jedním prvkem

GodsDev replied:
Doplnil jsem ještě inicializaci $original = null; a také && isset($command) do podmínky dále. Pokud vše proběhne ok, tak nebudou mít vliv. Pokud by něco selhalo, tak bude kód deterministický.

GodsDev commented on classes/MyTableAdmin.php:
```diff
@@ -511,13 +549,14 @@ public function recordSave($messageSuccess = false, $messageError = false)
         } else {
             Tools::addMessage('info', $this->translate('Nothing to save.'));
+            // todo ask CRS2  Method GodsDev\MyCMS\MyTableAdmin::recordSave() should return bool but returns int.
             return 0;
```
@crs2 metoda by měla vrátit bool, ale zde vrací 0. Co s tím?

crs2 replied:
Tady prosím pozměnit phpDoc u této metody (recordSave) na (něco jako):
/* @return bool|int - true = record saved sucessfully, false = error occured saving the record, 0 = no records to save
*/
důvod: ten, kdo tuhle metodu volá, by měl mít možnost rozlišit, jestli záznamy byly uloženy, nebo nebyly a nastala chyba, nebo odeslal k uložení 0 záznamů (předpokládám, že to je případ pro 0 zaškrtnutých checkboxů z formuláře).

Btw. v tom kódu je ještě @todo pro případ, kdy SQL proběhlo v pořádku, ale nic nebylo změněno (tj. byla správně uložena data tak, jak již byla v databázi). Je-li je kapacita to rovnou udělat, pak navrhuju přidat do recordSave() 3. parametr - &$affectedRows (s nějakou defaultní hodnotou, null třeba), do které se uloží ::$affected_rows (od třídy mysqli nebo potomka) v případě, že dojde až na vykonávání příkazu.

GodsDev replied:
Prima. To todo jsem přidal k metodě. V další fázi udělám. Slávek

GodsDev commented on classes/MyTableLister.php:
```diff
@@ -628,6 +638,7 @@ public function pagination($rowsPerPage, $totalRows, $offset = null, $options =
             }
             for ($page = 1; $page <= $pages; $page++) {
+                // TODO ask CRS2 Method GodsDev\MyCMS\MyTableLister::addPage() invoked with 6 parameters, 3-5 required.
                 $output .= $this->addPage($page, $currentPage, $rowsPerPage, null, $this->translate('Go to page'), $title);
```
@crs2 Ten 6.parametr $title je zřejmě zbytečně použit, neboť addPage akceptuje 3-5 parametrů?

crs2 replied:
Tady prosím vyhodit ", $this->translate('Go to page')" a nechat jako poslední dva parametry: null, $title

GodsDev commented on dist/classes/ProjectSpecific.php:
```diff
@@ -4,41 +4,49 @@
+            FROM ' . TAB_PREFIX . 'product WHERE product_' . $this->language
+            // TODO Expected at least 1 space after "+"; 0 found ask CRS2
+            . ' LIKE "%' . $q . '%" OR description_' . $this->language . ' LIKE "%' . $q . '%"
             LIMIT 10 OFFSET ' . +$offset);
```
@crs2 co dělá to + před $offset? Má to být int casting?

GodsDev reviewed (commented):
@crs2 Ahoj Karle, prošel jsem celý kód (vesměs pomocí PHPStan) a našel několik věcí z původně tvého kódu, se kterým si na první ani druhý pohled nevím rady. Mrkneš na to někdy pls? Dík Slávek

GodsDev committed:
Don't log SQL for user creation or password change (#38)

Security fix
* Don't log SQL for user creation or password change
* Log only INFO about the event of user creation or password change

Authored-by: Stanislav Rejthar, rejthar@stanislavrejthar.com

GodsDev commented on classes/MyTableAdmin.php:
```diff
@@ -223,57 +225,75 @@ protected function outputField(array $field, $key, array $record, array $options
             case 'date':
-                $input += [/*'type' => 'date',*/ 'class' => 'form-control input-date'];
+                // TODO fix Binary operation "+=" between arrays results in an error.
+                $input += [/* 'type' => 'date', */ 'class' => 'form-control input-date'];
```
@crs2 Co má dělat to +=  na typu array?

crs2 replied:
Po tom příkazu má být $input['class'] = 'form-control input-date'.
Dobrý postřeh - v tomhle případě je chybné pořadí:
správně by mělo být: $input = ['class' => 'form-control input-date'] + $input;
Pro jednoduchost a lepší čitelnost bych byl proto, aby tam byla prostá přiřazování do pole, v tomhle případě:
$input['class'] = 'form-control input-date';
a u ostatních (i když je více prvků) tak každý na samostatné řádce.

GodsDev replied:
Aha. A + mezi array je union. Good. Slávek

GodsDev committed:
neither admin table nor its rows can be accessed through admin.php (#39)

GodsDev committed:
replaces older way of type casting `+$int` by new standard way `(int) $int`

GodsDev committed:
Cosmetic changes learned from Stockpiler project

GodsDev committed:
PHPUnit test by GitHub actions (#41)

 PHPUnit test by GitHub actions
* Skip dist tests until MySQLi is prepared
* Btw:  HTTP requests to self can't work in CLI only environment

Governance
* class path moved to `use` statement
* FaviconTest.php checks for presence of favicon resources
* workflow test badges

GodsDev committed:
Decluttering favicons from dist folder (#42)

* Favicon resources moved to images/favicon folder, but due to RewriteRule in [.htaccess](dist/.htaccess) they still seem to be present where devices are automatically looking for them
* PHPUnit test automatically checks the availability of those resources

GodsDev committed:
* MyAdmin: function json_encode shouldn't use #2 parameter
* MyTableLister: method addPage has just 5 parameters
* AdminProcess: use TableAdmin (instead of MyTableAdmin) so that $this->tableAdmin->customDelete() refers to an existing method
* MyTableAdmin: recordSave() $field instead of $value; check for $this->fields being be an array with at least one element
* If you want to append array elements from the second array to the first array while not overwriting the elements from the first array and not re-indexing, use the + array union operator: fix Binary operation "+=" between arrays results in an error ==> for simplicity and better readability use simple assignment, i.e.
`$input['class'] = 'form-control input-date';` instead of `$input += [/* 'type' => 'date', */ 'class' => 'form-control input-date'];` which should be anyway `$input = ['class' => 'form-control input-date'] + $input;`
* AdminProcess: method adminProcess() fix Binary operation "+=" between array&nonEmpty and array|false results in an error.

GodsDev commented on classes/MyAdminProcess.php:
```diff
@@ -70,10 +73,11 @@ protected function filterChecks($checks, &$errors)
             foreach (explode('&', $check) as $condition) {
                 $condition = explode('=', $condition, 2);
                 if (Tools::begins($condition[0], 'where[') && Tools::ends($condition[0], ']')) { //@todo doesn't work for nulls
+                    // TODO fix: Call to function is_null() with string will always evaluate to false.
```
@crs2 výsledkem explode budou strings, má tedy smysl na nich testovat is_null ? Slávek

GodsDev commented on classes/MyTableAdmin.php:
```diff
@@ -336,20 +369,21 @@ protected function outputField(array $field, $key, array $record, array $options
      */
     public function outputSelectPath($name, $path_id = null, $options = [])
     {
+        // TODO what does this construction mean? Call to function is_array() with string will always evaluate to false.
```
@crs2 Znamená to, že buď je $name array, kde definuji pole table a column anebo je to string, a pak se má pracovat s polem, kde table i column je tím řetězcem? Slávek

GodsDev commented on classes/MyTableAdmin.php:
```diff
@@ -476,31 +521,35 @@ public function recordSave($messageSuccess = false, $messageError = false)
             }
             $command = 'UPDATE';
+            // todo fix Parameter #1 $types of method GodsDev\MyCMS\MyTableLister::filterKeys() expects array, string given.
             $unique = ($this->filterKeys('PRI') ?: $this->filterKeys('UNI')) ?: array_flip(array_keys($this->fields));
```
@crs2 filterKeys tedy může použít jako parametr nejen array, ale legitimně i string? Nebylo by čitelnější upravit volání `$this->filterKeys('PRI')` na `$this->filterKeys(['PRI'])` a metodu filterKeys předělat, aby akceptovala jen array?

GodsDev commented on classes/MyTableLister.php:
```diff
@@ -338,6 +348,7 @@ public function bulkUpdateSQL(&$vars)
                     break;
             }
         }
+        // todo fix Method GodsDev\MyCMS\MyTableLister::bulkUpdateSQL() should return string but return statement is missing.
```
@crs2 buď by tam mělo být `return $result;` anebo metoda nemá vracet nic?

GodsDev commented on classes/MyTableLister.php:
```diff
@@ -945,6 +957,7 @@ public function contentByType(array $options = [])
             . ' FROM ' . Tools::escapeDbIdentifier(TAB_PREFIX . $options['table'])
             . ' GROUP BY ' . Tools::escapeDbIdentifier($options['type']) . ' WITH ROLLUP LIMIT 100');
         if (!$query) {
+            // TODO fix Method GodsDev\MyCMS\MyTableLister::contentByType() should return string but empty return statement found.
```
@crs2  má zde být `return null;` nebo `return '';` ? Slávek

GodsDev commented on classes/MyTableLister.php:
```diff
@@ -967,6 +980,7 @@ public function contentByType(array $options = [])
             return $output;
         }
         echo $output;
+        // TODO fix Method GodsDev\MyCMS\MyTableLister::contentByType() should return string but return statement is missing.
```
@crs2 Nemají 4 poslední řádky být spíš
```php
        if (!(isset($options['return-output']) && $options['return-output'])) {
            echo $output;
        }
        return $output;
```
...? Aby return string.

GodsDev committed:
MyTableLister::rowLink() fix Strict comparison using === between mixed and null will always evaluate to false

GodsDev reviewed (commented):
Přidal jsem ještě 6 drobností. Slávek
